File: intelligent.py
import math


def compute_primes(x):
    length = int(math.sqrt(x)) +1
    l = [False for i in range(length-1)]
    
    for ind,i in enumerate(l):
        if l[ind] or ind==0:
            continue
        numb = ind+1
        result = x/numb
        if result == int(result):
            return result, numb
        else:
            l[ind]= True
            k = numb
            while k*numb <= length-1 and numb<10:
                l[k*numb-1]= True
                k += numb

File: intelligent.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

void compute_primes(long long x, long long *result, long long *factor) {
    if (x % 2 == 0) {
        *result = x / 2;
        *factor = 2;
        return;
    }

    long long length = (long long)sqrt((double)x) + 1;
    for (long long i = 3; i <= length; i += 2) {
        if (x % i == 0) {
            *result = x / i;
            *factor = i;
            return;
        }
    }

    *result = x; // x is prime
    *factor = 1;
}
